torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+    to follow the same order and have the same output tensor livespans. To remove
+    unnecessary coupling of cuda graphs (and additional imposed invariants),
+    the tree manager will end a currently recording tree whenever it is valid - when
+    the memory pool no longer has any live allocations.
+
+    We ignore outputs from a previous generation that correspond to prior model outputs.
+    Currently this is hardcoded `GenerationTracker.generation` tracked in torch dynamo.
+    # TODO: make generation increment configurable, warn on overwrite
+    """
+
+    def __init__(self):
+        self.roots: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
+        self.ids_to_funcs: Dict[FunctionID, WrappedFunction] = {}
+
+        self.cuda_graphs_thread_pool = torch.cuda.graph_pool_handle()
+
+        self.graph_counter = itertools.count(0)
+        self.func_counter = itertools.count(0)
+
+        self.current_node: Optional[CUDAGraphNode] = None
+        self.current_gen: int = -1
+
+        self.in_recording = False
+        self.previous_recording_outputs: List[TensorWeakRef] = []
+
+        self.debug_fail_counter = 0
+        self.debug_checkpointing_counter = 0
+
+    def run(self, new_inputs: List[Tensor], function_id: FunctionID):
+        # we will try to end the current execution lazily, since
+        # we dont want to do unnecessary checking of the existing outputs
+        # on the hot path
+        if self.in_recording:
+            self.try_end_curr_recording()
+
+        child_nodes = (
+            self.roots if self.current_node is None else self.current_node.children
+        )
+
+        if not self.in_recording:
+            for child in child_nodes[function_id]:
+                # here we are checking memory consistency between recording and execution,
+                # as well as things like stability of tensor locations, etc
+                # and other
+                if child.check_invariants(new_inputs):
+                    self.current_gen = self.get_curr_generation()
+                    return self.execute_node(child, new_inputs)
+
+            # now that we know the new function can't be run as a child of the
+            # current node, if it is a root, try to end the current execution.
+            # as noted above, we want to do this lazily to avoid having to
+            # check all existing outputs
+            if self.current_node is not None and function_id in self.roots:
+                self.try_end_curr_execution()
+
+                # run again to hit the root matching case which must succeed
+                if self.current_node is None:
+                    return self.run(new_inputs, function_id)
+
+            # at this point, we necessarily will do a new recording
+            self.debug_fail_counter += 1
+
+            self.try_end_curr_execution()
+            if self.current_node is not None:
+                self.convert_execution_to_recording()
```

Comment:
dumb Q: where did the execution here happen?

Reply:
I renamed the function, does that clarify things?